File: clients/android/app/src/main/java/asia/tnestai/vpn/MultiRegionProfile.kt
package asia.tnestai.vpn

import org.json.JSONArray
import org.json.JSONObject

data class CatalogNode(
    val id: String,
    val endpoint: String,
    val probeUrl: String,
    val serverPublicKey: String,
    val priority: Int
) {
    fun toJson() = JSONObject()
        .put("id", id).put("endpoint", endpoint).put("probe_url", probeUrl)
        .put("server_public_key", serverPublicKey).put("priority", priority)

    companion object {
        fun fromJson(value: JSONObject) = CatalogNode(
            value.getString("id"), value.getString("endpoint"),
            value.getString("probe_url"), value.getString("server_public_key"),
            value.getInt("priority")
        )
    }
}

data class RegionConfiguration(
    val code: String,
    val displayName: String,
    val addresses: List<String>,
    val dns: List<String>,
    val mtu: Int,
    val configVersion: Int,
    val nodes: List<CatalogNode>,
    val privateKey: String,
    val presharedKey: String
) {
    fun render(node: CatalogNode): String {
        require(nodes.any { it.id == node.id })
        return """
            [Interface]
            PrivateKey = $privateKey
            Address = ${addresses.joinToString(", ")}
            DNS = ${dns.joinToString(", ")}
            MTU = $mtu

            [Peer]
            PublicKey = ${node.serverPublicKey}
            PresharedKey = $presharedKey
            Endpoint = ${node.endpoint}
            AllowedIPs = 0.0.0.0/0, ::/0
            PersistentKeepalive = 25
        """.trimIndent() + "\n"
    }

    fun toJson() = JSONObject()
        .put("code", code).put("display_name", displayName)
        .put("address", JSONArray(addresses)).put("dns", JSONArray(dns))
        .put("mtu", mtu).put("config_version", configVersion)
        .put("nodes", JSONArray(nodes.map(CatalogNode::toJson)))
        .put("private_key", privateKey).put("preshared_key", presharedKey)

    companion object {
        fun fromJson(value: JSONObject) = RegionConfiguration(
            value.getString("code"), value.getString("display_name"),
            value.getJSONArray("address").strings(),
            value.getJSONArray("dns").strings(), value.getInt("mtu"),
            value.getInt("config_version"),
            value.getJSONArray("nodes").objects().map(CatalogNode::fromJson),
            value.getString("private_key"), value.getString("preshared_key")
        )
    }
}

data class MultiRegionProfile(
    val deviceName: String,
    val managementUrl: String,
    val deviceToken: String,
    val catalogSigningKey: String,
    val catalogVersion: Long,
    val catalogExpiresAt: String,
    val regions: List<RegionConfiguration>
) {
    fun region(code: String) = regions.first { it.code.equals(code, ignoreCase = true) }

    fun toJson() = JSONObject()
        .put("version", 2).put("device_name", deviceName)
        .put("management_url", managementUrl).put("device_token", deviceToken)
        .put("catalog_signing_key", catalogSigningKey)
        .put("catalog_version", catalogVersion).put("catalog_expires_at", catalogExpiresAt)
        .put("regions", JSONArray(regions.map(RegionConfiguration::toJson))).toString()

    companion object {
        fun fromJson(text: String): MultiRegionProfile {
            val value = JSONObject(text)
            require(value.getInt("version") == 2)
            return MultiRegionProfile(
                value.getString("device_name"), value.getString("management_url"),
                value.getString("device_token"), value.getString("catalog_signing_key"),
                value.getLong("catalog_version"), value.getString("catalog_expires_at"),
                value.getJSONArray("regions").objects().map(RegionConfiguration::fromJson)
            )
        }
    }
}

private fun JSONArray.strings() = (0 until length()).map(::getString)
private fun JSONArray.objects() = (0 until length()).map(::getJSONObject)

File: clients/windows/src/TNestVPN.Core/MultiRegionProfile.cs
using System.Text.Json.Serialization;

namespace TNestVPN.Core;

public sealed record CatalogNode(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("endpoint")] string Endpoint,
    [property: JsonPropertyName("probe_url")] string ProbeUrl,
    [property: JsonPropertyName("server_public_key")] string ServerPublicKey,
    [property: JsonPropertyName("priority")] int Priority);

public sealed record CatalogRegion(
    [property: JsonPropertyName("code")] string Code,
    [property: JsonPropertyName("display_name")] string DisplayName,
    [property: JsonPropertyName("sort_order")] int SortOrder,
    [property: JsonPropertyName("exit_mode")] string ExitMode,
    [property: JsonPropertyName("config_version")] int ConfigVersion,
    [property: JsonPropertyName("nodes")] IReadOnlyList<CatalogNode> Nodes);

public sealed record SignedCatalog(
    [property: JsonPropertyName("catalog_version")] long Version,
    [property: JsonPropertyName("issued_at")] DateTimeOffset IssuedAt,
    [property: JsonPropertyName("expires_at")] DateTimeOffset ExpiresAt,
    [property: JsonPropertyName("regions")] IReadOnlyList<CatalogRegion> Regions,
    [property: JsonPropertyName("signature")] string Signature,
    [property: JsonPropertyName("signed_payload")] string SignedPayload);

public sealed record RegionConfiguration(
    [property: JsonPropertyName("region_code")] string RegionCode,
    [property: JsonPropertyName("address")] IReadOnlyList<string> Address,
    [property: JsonPropertyName("dns")] IReadOnlyList<string> Dns,
    [property: JsonPropertyName("mtu")] int Mtu,
    [property: JsonPropertyName("allowed_ips")] IReadOnlyList<string> AllowedIps,
    [property: JsonPropertyName("persistent_keepalive")] int PersistentKeepalive,
    [property: JsonPropertyName("config_version")] int ConfigVersion,
    [property: JsonPropertyName("nodes")] IReadOnlyList<CatalogNode> Nodes);

public sealed record RegionSecret(
    string RegionCode,
    string PrivateKey,
    string PresharedKey,
    RegionConfiguration Configuration);

public sealed record MultiRegionProfile(
    string DeviceName,
    Uri ManagementUrl,
    string DeviceToken,
    string CatalogSigningKey,
    SignedCatalog Catalog,
    IReadOnlyList<RegionSecret> Regions)
{
    public RegionSecret Region(string code) =>
        Regions.Single(region => string.Equals(
            region.RegionCode, code, StringComparison.OrdinalIgnoreCase));

    public WireGuardConfig Render(string regionCode, CatalogNode node)
    {
        var region = Region(regionCode);
        if (!region.Configuration.Nodes.Any(candidate => candidate.Id == node.Id))
            throw new InvalidOperationException("节点不属于所选地区");
        var text = $"""
            [Interface]
            PrivateKey = {region.PrivateKey}
            Address = {string.Join(", ", region.Configuration.Address)}
            DNS = {string.Join(", ", region.Configuration.Dns)}
            MTU = {region.Configuration.Mtu}

            [Peer]
            PublicKey = {node.ServerPublicKey}
            PresharedKey = {region.PresharedKey}
            Endpoint = {node.Endpoint}
            AllowedIPs = 0.0.0.0/0, ::/0
            PersistentKeepalive = 25

            """;
        return SafeWireGuardConfig.Parse(text);
    }
}
